src/44bsd/tcp_socket.cpp

```diff
@@ -1100,19 +1100,48 @@ int utl_mbuf_buffer_create_and_copy(uint8_t socket,
 }
 
 
+/*
+ * The mbuf filled with patterns can be cached for other buffers.
+ * It will reduce 2K mbuf usage when lots of buffers are loaded.
+ * To avoid the 16bit refcnt overflow, const-mbuf is used.
+ * The const-mbuf could not be freed because it will stay in the driver's TX descriptor.
+ * In general, the number of cached mbufs is small and has no issue.
+ */
+static std::map<std::string,rte_mbuf_t*> mbufs_cached[MAX_SOCKETS_SUPPORTED];
+static std::mutex mbufs_cached_mutex[MAX_SOCKETS_SUPPORTED];
+
+static rte_mbuf_t* get_mbuf_cached(uint8_t socket, std::string& fstring) {
+    std::unique_lock<std::mutex> my_lock(mbufs_cached_mutex[socket]);
+
+    auto it = mbufs_cached[socket].find(fstring);
+
+    return (it != mbufs_cached[socket].end()) ? it->second: nullptr;
+}
+
+static void set_mbuf_cached(uint8_t socket, std::string& fstring, rte_mbuf_t* mbuf) {
+    if (CGlobalInfo::m_do_mbuf_cache) {
+        std::unique_lock<std::mutex> my_lock(mbufs_cached_mutex[socket]);
+        assert(mbufs_cached[socket].find(fstring) == mbufs_cached[socket].end());
+
+        mbufs_cached[socket][fstring] = mbuf;
+        rte_mbuf_set_as_core_const(mbuf);
+    }
+}
+
+
 int utl_mbuf_buffer_create_and_copy(uint8_t socket,
                                     CMbufBuffer * buf,
                                     uint32_t blk_size,
                                     uint8_t *d,
                                     uint32_t d_size,
                                     uint32_t size,
-                                    uint8_t *f,
-                                    uint32_t f_size,
+                                    std::string& fill_string,
                                     bool mbuf_const){
 
     buf->Create(blk_size);
     uint8_t cnt = 0;
-    rte_mbuf_t* mbuf_fill = nullptr;
+    rte_mbuf_t* mbuf_fill = get_mbuf_cached(socket, fill_string);
```

Comment:
if it was found, it is the same string in the mbuf and you can skip the rest of the function, you could return here

Reply:
@jsmoon right, you are correct. The return object is of another type.